shared/src/main/scala/async/retry.scala

```diff
@@ -0,0 +1,163 @@
+          if retryOnSuccess then
+            sleep(delay.delayFor(0, 0.second).toMillis)
+            loop(0, 0.second)
+          else value
+
+    loop(0, 0.seconds)
+
+  /** Set the maximum failure count. */
+  def withMaximumFailures(max: Int) =
+    assert(max >= 0)
+    this.copy(maximumFailures = Some(max))
+
+  /** Set the delay policy between runs. See [[Delay]]. */
+  def withDelay(delay: Delay) = this.copy(delay = delay)
+
+object Retry:
+  /** Ignores the result and attempt the action in an infinite loop. [[Retry.withMaximumFailures]] can be useful for
+    * bailing on multiple failures. [[scala.util.boundary]] can be used for manually breaking.
+    */
+  val forever = Retry(retryOnSuccess = true)
+
+  /** Returns the result, or attempt to retry if an exception is raised. */
+  val untilSuccess = Retry(retryOnSuccess = false)
+
+  /** Attempt to retry the operation *until* an exception is raised. In this mode, [[Retry]] always throws an exception
+    * on return.
+    */
+  val untilFailure = Retry(retryOnSuccess = true).withMaximumFailures(0)
+
+  /** Defines a delay policy based on the number of successive failures and the duration of the last delay. See
+    * [[Delay]] companion object for some provided delay policies.
+    */
+  trait Delay:
+    /** Return the expected duration to delay the next attempt from the current attempt.
+      *
+      * @param failuresCount
+      *   The number of successive failures until the current attempt. Note that if the last attempt was a success,
+      *   [[failuresCount]] is `0`.
+      * @param lastDelay
+      *   The duration of the last delay, if there were any successive failures. Otherwise this is `0`.
+      */
+    def delayFor(failuresCount: Int, lastDelay: Duration): Duration
+
+  object Delay:
+    /** No delays. */
+    val none = constant(0.second)
+
+    /** A fixed amount of delays, whether the last attempt was a success or failure. */
+    def constant(duration: Duration) = new Delay:
+      def delayFor(failuresCount: Int, lastDelay: Duration): Duration = duration
+
+    /** Returns a delay policy for exponential backoff.
+      * @param maximum
+      *   The maximum duration possible for a delay.
+      * @param starting
+      *   The delay duration between successful attempts, and after the first failures.
+      * @param multiplier
+      *   Scale the delay duration by this multiplier for each successive failure. Defaults to `2`.
+      * @param jitter
+      *   An additional jitter to randomize the delay duration. Defaults to none. See [[Jitter]].
+      */
+    def backoff(maximum: Duration, starting: Duration, multiplier: Double = 2, jitter: Jitter = Jitter.none) =
+      new Delay:
+        def delayFor(failuresCount: Int, lastDelay: Duration): Duration =
+          jitter
+            .jitterDelay(
+              lastDelay,
+              if failuresCount <= 1 then starting
+              else (lastDelay.toMillis * multiplier).toLong.millis
```

Comment:
Shouldn't this rather be $multiplier^{failuresCount-1} * starting$ to work well with jittering? Otherwise, e.g., for multiplier=2 and full jitter, we will end up with an effectively constant delay (expectation value of actual delay cancels out factor 2).